### PR TITLE
🎉 Source Hubspot: Added custom object streams

### PR DESCRIPTION
**Pull request by lgomezm**
## What
* Husbpot custom objects are not currently supported as streams.
* Their schemas can't be added as fixed files, because they change from account to account.

## How
* When computing available streams, it checks if it can get custom object schemas to add them as available streams.
* Extends the `CrmSearchStream` to add custom object streams.
* It also adds the `crm.objects.deals.read` scope to the `DealPipelines` stream, as it can also be used to get data from that stream.

## Recommended reading order
1. `source.py`
2. `streams.py`
3. `test_streams.py`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>
Screen Shot 2022-08-26 at 5 53 02 PM
Screen Shot 2022-08-26 at 5 54 16 PM

Screen Shot 2022-08-26 at 5 54 59 PM
Screen Shot 2022-08-26 at 5 55 12 PM

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by sajarin** [comment]
Hey @lgomezm, this PR is part of the Community Maintainers program. Someone from our community will be assigned to review your PR and help it get merged into our main branch. Thanks for your contribution and we appreciate your patience.

**Comment by sajarin** [comment]
assigned to @tuanchris

**Review by tuanchris — COMMENTED** [review]
Hi @lgomezm, just got the Hubspot credentials from the Airbyte team today, so sorry for the delay. I was able to run unit tests and integration tests (for other streams) successfully on my local machine. 

For the `CustomObject` stream, since the test account doesn't have any custom object, I was not able to test that. Can you post a passing screenshot/logs for that particular stream? 

Everything else looks good.

**Comment by tuanchris** [comment]
@lgomezm Just following up on this. Can I help with anything?

**Comment by gfrts** [comment]
Hi @tuanchris, how's it going? I'm very interested in this PR, is a helpful feature and solve some problems that i have been faced recently. Can i help with Hubspot and create a custom object  in this airbyte account to allow the test conclusion? That's a possibility?
I appreciate the effort of all here and i wanna help!

**Comment by tuanchris** [comment]
Hi @gfrts, thank you for jumping in! If you can help test out the new `CustomObject` stream, we can look to merge this PR!

**Comment by gfrts** [comment]
@tuanchris nice! I'm pretty new to the community, I still need to install airbyte locally for testing. I must do this today.
As for hubspot, I can use any account that has custom objects to test the stream, right?

**Comment by tuanchris** [comment]
@gfrts welcome! This is a great place to start: https://docs.airbyte.com/contributing-to-airbyte/developing-locally/. After you set up everything, you can check out this branch. You can use any hubspot acc witth custom objects.

**Comment by gfrts** [comment]
hey @tuanchris! After suffering a bit, i managed to get airbyte to run locally, from this tutorial https://docs.airbyte.com/deploying-airbyte/local-deployment/#deploy-on-windows (I use windows). To be able to test this new feature, I forked the repository and made the changes from this pullrequest, and then ran git clone on my repo.
When I tried to create the hubspot source to test, he stayed on this screen for hours (see image), and I don't know how to proceed, unfortunately. Any idea?

image

**Comment by natalyjazzviolin** [comment]
@gfrts if the screen freezes like that, try opening developer tools in that tab and see if there are errors in the console log!

**Comment by tuanchris** [comment]
@gfrts Sorry for the late reply. It looks like you also are using the official Hubspot image instead of the one built from this branch. If you don't mind sharing Hubspot credentials via Slack, I'd be happy to have a look at it. 

Otherwise, you can run `docker build . -t airbyte/source-hubspot:dev` in the connector directory, and add a custom connector following the instructions [here](https://docs.airbyte.com/integrations/custom-connectors/). Be sure to have the tag as `dev`.

**Comment by gfrts** [comment]
@tuanchris and @natalyjazzviolin, thanks for the feedback and sorry for the delay! This week has been busy, and i will need to allocate a few hours to retest (i'm new to deployments and all that stuff). I should work on it this weekend.

@tuanchris unfortunately the hubspot account I'm using has sensitive data, so I can't share it :/

**Comment by lgomezm** [comment]
@tuanchris Hi! I am sorry I was out recently. These are screenshots of an execution of the hubspot connector with the new custom object stream. In this example, I am reading records of a custom object called `team` in the test account. I hope this helps get this PR merged. Please let me know if I can do anything else.

Screen Shot 2022-10-10 at 10 37 54 AM
Screen Shot 2022-10-10 at 10 39 20 AM

**Comment by lgomezm** [comment]
@tuanchris I ran those tests after I rebased my branch, so that it ran with the most up-to-date code (since the PR has been open for a few weeks 🙂).

**Review by tuanchris — APPROVED** [review]
Great work! Thanks, @lgomezm.
CC @sajarin

**Comment by gfrts** [comment]
@tuanchris @natalyjazzviolin @lgomezm I'm not very experienced with git, I saw that the merge was not done yet because of an error (_Report connectors dependency / Report affected connectors (pull_request)_ ), but I didn't understand what it means. Do you know what's going on?

**Comment by lgomezm** [comment]
@gfrts The flake check had failed. 
```
> Task :airbyte-integrations:connectors:source-hubspot:flakeCheck FAILED
[python] .venv/bin/python -m pflake8 --config /actions-runner/_work/airbyte/airbyte/pyproject.toml ./
	 ./unit_tests/test_streams.py:296:5: F841 local variable 'expected_schema' is assigned to but never used
```
I've just updated the code

**Comment by sajarin** [comment]
@lgomezm the tests are failing due to a new cdk version. **Can you go to setup.py and update the airbyte-cdk version from 0.1.55 to 0.2.0** 

Also can you bump the dockerfile and update the readme.md?

**Comment by lgomezm** [comment]
@sajarin I've just updated the cdk version and rebased. Hopefully this is it!

**Comment by luizmcarvalho** [comment]
Hi @lgomezm. I'm looking forward to this feature of Custom Objects.

What does it take to move forward? Thank you for this great job!

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by lgomezm** [comment]
@luizmcarvalho @marcosmarxm I'm sorry I've been dragged to other things and haven't payed much attention to this PR. I've just rebased, because this branch had some merge conflicts. I am not sure why the acceptance tests didn't pass last time (maybe it has to do with new CDK?). Could they be triggered again?

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by AlexandreKeo** [comment]
Hi @lgomezm  ! Very interested by this feature also (actually, it's a mandatory feature for me to be able to suggest Airbyte as  our main Ingestion tool in my company) => Do you know when it will be merged ? (I see that the tests seem hard to pass)
Thank you for your work by the way, I'm sure it will be useful for a lots of people out there !

**Comment by zkid18** [comment]
Hey @lgomezm looking forward to the feature as well. Is there anything I can help with?

**Comment by lgomezm** [comment]
Hey guys. I'm sorry I had left this PR unattended for so long. I have just rebased it off of master and ran both unit and integration tests locally. Both work, as it can be seen below:
**Unit tests:**
Screenshot 2023-03-06 at 7 14 27 AM
Screenshot 2023-03-06 at 7 09 42 AM
Screenshot 2023-03-06 at 7 10 05 AM
Screenshot 2023-03-06 at 7 10 26 AM

**Integration tests:**
Screenshot 2023-03-06 at 7 25 14 AM
Screenshot 2023-03-06 at 7 25 34 AM

@marcosmarxm What can I do to get this PR merged soon?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py
Comment (igsaf2) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:132]: It seems that your feature is not supported if an API Key used for authentication. Is it intentional?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (igsaf2) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:204]: It misses the `type` attribute. Some destination connectors can crash because of that.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (igsaf2) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:223]: I am not sure if it is important, but the schema can be empty here (e.g. on `bool`)

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py]: Good catch. Added that in [eb427b1](https://github.com/airbytehq/airbyte/pull/16032/commits/eb427b19ca401c76cabc66290bb21bf23b609f51).

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py
Comment (lgomezm) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:132]: I think Hubspot is deprecating API Keys, so I don't know if this can be supported with that. My personal hubspot account can't create any new API Key to test that, so I don't know if I should allow that.

At my company, we have been running a custom version of this connector for a while and it has worked with OAuth tokens.

**Comment by marcosmarxm** [comment]
I added this to my to-do list @lgomezm

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1305]: According to docs we don't need deals scopes, only contacts for deals pipeline or tickets - https://legacydocs.hubspot.com/docs/methods/pipelines/get_pipelines_for_object_type

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:121]: nice

**Review by lazebnyi — CHANGES_REQUESTED** [review]
Need change scopes and pass tests. After that I will approve. Other stuff LGTM!

**Review by lazebnyi — APPROVED** [review]

**Comment by apd-wguicheney** [comment]
Hey team!

Any timeline on when it will be possible to pull custom objects using Airbyte?

**Comment by marcosmarxm** [comment]
Waiting to update Hubspot OAuth Scope to merge this contribution.

**Comment by philk1991** [comment]
Any update on this one? Would be super useful to be able to load custom objects

**Review by marcosmarxm — APPROVED** [review]

**Comment by apd-wguicheney** [comment]
Hey team! 

I'm seeing a lot of updates here - does this seem that we are now able to pull custom objects?

Thank you!!

**Pull request merged**
